### ISSUE TITLE
Dose min-max hinge loss right?

### ISSUE THREAD

**Issue by syzlhh** [salu133445/dan#1]
def minimax_hinge_gan_losses(discriminator_real_outputs,
                             discriminator_fake_outputs):
    """Return the minimax hinge GAN losses for the generator and the
    discriminator.
    **(Generator)      -max(0, 1 + D(G(z)))
    (Discriminator)  max(0, 1 - D(x)) + max(0, 1 + D(G(z)))**
    """
    discriminator_loss = (
        tf.reduce_mean(tf.nn.relu(1. - discriminator_real_outputs))
        + tf.reduce_mean(tf.nn.relu(1. + discriminator_fake_outputs)))
    generator_loss = -discriminator_loss
    return generator_loss, discriminator_loss

the  note is not the same as the code in the function, Can you give an explanation? Thanks.

**Comment by salu133445** [comment]
For simplicity, the code is written using the ReLU function. As the ReLU function is by definition `relu(x) = max(0, x)`, it actually implements the formulas presented in the note.

**Comment by syzlhh** [comment]
Thanks for your answer. But the generator loss is the reverse of discriminator loss, which contain the max(0,1-D(x)) in the code. And the note doesn't add this item.

**Comment by salu133445** [comment]
Since the term `max(0, 1-D(x))` has no gradients w.r.t the generator parameters, it does not change the overall training objective. In other words, it is a constant term w.r.t to the generator.
